File: entity-meta-generator-lib/src/main/java/com/loudsight/meta/serialization/transform/CollectionEntityTransform.java
package com.loudsight.meta.serialization.transform;

import com.loudsight.meta.serialization.EntityTransform;
import com.loudsight.meta.serialization.EntityTransforms;
import com.loudsight.meta.serialization.EntityType;

import java.util.Collection;
import java.util.Iterator;
import java.util.List;
import java.util.function.Function;

public abstract class CollectionEntityTransform<T extends Collection<?>> extends EntityTransform<T> {
    public CollectionEntityTransform(EntityType entityType, Class<T> kClass) {
        super(entityType, kClass);
    }

    abstract T newCollection(int size, Function<Integer, Object> init);

    @Override public void serializeEntity( T entity, List<Byte> bytes) {
        bytes.add(entityType.getCode());
        writeInt(entity.size(), bytes);

        entity.forEach(it -> {
            if (it == null) {
                bytes.add(EntityType.NULL.getCode());
            } else {
                EntityTransform<Object> entityTransform = EntityTransforms.getInstance().getEntityTransform(it);
                entityTransform.serializeEntity(it, bytes);
            }
        });
    }

    @Override public T  deserializeEntity(Iterator<Byte> bytes) {
        var size = readInt(bytes);

        return newCollection(size, it -> {
            var entityType1 = EntityType.getEntityType(bytes.next());
            var transform = EntityTransforms.getInstance().getEntityTransform(entityType1);
            return transform.deserializeEntity(bytes);
        });
    }
}

File: entity-meta-generator-lib/src/main/kotlin/com/loudsight/meta/serialization/transform/CollectionEntityTransform.kt
package com.loudsight.meta.serialization.transform

import com.loudsight.meta.serialization.EntityTransform
import com.loudsight.meta.serialization.EntityTransforms
import com.loudsight.meta.serialization.EntityType
import kotlin.reflect.KClass

abstract class CollectionEntityTransform<T: Collection<*>>(entityType: EntityType, entityClass: KClass<T>)
    : EntityTransform<T>(entityType, entityClass) {

    abstract fun newCollection(size: Int, init: (index: Int) -> Any?): T

    override fun serializeEntity(entity: T, bytes: MutableList<Byte>) {
        bytes.add(entityType.code)
        writeInt(entity.size, bytes)

        entity.forEach {
            if (it == null) {
                bytes.add(EntityType.NULL.code)
            } else {
                val entityTransform: EntityTransform<Any>? = EntityTransforms.getEntityTransform<Any>(it)
                entityTransform?.serializeEntity(it, bytes)
            }
        }
    }

    override fun deserializeEntity(bytes: Iterator<Byte>): T {
        val size = readInt(bytes)

        val collection = newCollection(size)
        {
            val entityType = EntityType.getEntityType(bytes.next().toInt().toChar())
            val transform = EntityTransforms.getEntityTransform<Any>(entityType)
            transform?.deserializeEntity(bytes)
        }

        return collection
    }
}